### PR TITLE
[R] Allow sorting on donor count (#1378)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**

### CODE REVIEW

File: src/azul/transformer.py [hunk 1]
```diff
@@ -648,7 +648,7 @@ def accumulate(self, value) -> bool:
         """
         :return: True, if the given value increased the count of unique values
         """
-        return self.items.accumulate(value)
+        return self.set_accum.accumulate(value)
```
Comment (jessebrennan) [inline src/azul/transformer.py]: This change belongs in the previous commit. If you'd rather not move the change, consider squashing these together.

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 2]
```diff
@@ -822,7 +822,14 @@ def make_donor(self, donor):
...
     def make_donors(self, entry):
-        return [self.make_donor(donor) for donor in entry["contents"]["donors"]]
+        donors = entry["contents"]["donors"]
+        if donors:
+            count = donors[0]['donor_count']
+            result = [self.make_donor(donor) for donor in donors]
+            result[0]['donorCount'] = count
+            return result
+        else:
+            return []
```
Comment (jessebrennan) [inline src/azul/service/responseobjects/hca_response_v5.py]: I'm confused why this change made here. I think instead you should be changing the `self.make_donor()` function.

Comment (jessebrennan) [reply src/azul/service/responseobjects/hca_response_v5.py]: I'll leave this to @hannes-ucsc's discretion.

Comment (hannes-ucsc) [reply src/azul/service/responseobjects/hca_response_v5.py]: Unless I'm mistaken, `len(entry["contents"]["donors"]) < 2` invariantly. If, in the future, we started to group donors by some property (like we do for files which are grouped by format), then there may be more than one donor in an aggregate document. In that case we'd want a count in every donor in the response. My instructions where imprecise in that respect. I think we should just modify make_donor as @jessebrennan suggests, and have the count in every donor in the response.

File: src/azul/transformer.py (hunk 1 above)
Comment (hannes-ucsc) [reply src/azul/transformer.py]: By our current, ill-defined standards, the history is a bit fragmented. All of the commits are tightly coupled, so much so that I don't foresee us reverting one of them and leaving the others. That's my goto criterion for deciding what goes into a commit: would I ever want to revert a change separately from another change. If yes, separate commit. Otherwise, same commit.

File: src/azul/transformer.py [hunk 3]
```diff
@@ -635,6 +635,26 @@ def get(self):
         return self.value.get()
 
 
+class SetCountingAccumulator(Accumulator):
```
Comment (hannes-ucsc) [inline src/azul/transformer.py]:
```suggestion
class CountUniqueValuesAccumulator(Accumulator):
```

File: src/azul/transformer.py [hunk 4]
```diff
@@ -635,6 +635,26 @@ def get(self):
...
+    """
+    Count the number of unique values
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.set_accum = SetAccumulator()
```
Comment (hannes-ucsc) [inline src/azul/transformer.py]:
```suggestion
        self.value = SetAccumulator()
```

For consistency with other accumulators.

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 5]
```diff
@@ -814,6 +814,7 @@ def make_cell_lines(self, entry):
     def make_donor(self, donor):
         return {
             "id": donor["biomaterial_id"],
+            "donorCount": donor["donor_count"],
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: The IT fails when hitting the service's health endpoint which in turns hits the repository endpoints. Hitting those manually after enabling AZUL_DEBUG yields.

```
$ http 'https://service.sandbox.dev.explore.data.humancellatlas.org/repository/bundles?size=1'
HTTP/1.1 500 Internal Server Error

Traceback (most recent call last):
  File "/var/task/chalice/app.py", line 726, in _get_view_function_response
    response = view_function(**function_args)
  File "/var/task/app.py", line 580, in get_bundle_data
    return repository_search('bundles', bundle_uuid)
  File "/var/task/app.py", line 435, in repository_search
    return service.get_data(entity_type, pagination, filters, item_id, file_url)
  File "/var/task/azul/service/repository.py", line 86, in get_data
    return self._get_items(entity_type, pagination, filters, file_url_func)
  File "/var/task/azul/service/repository.py", line 60, in _get_items
    response = self._get_data(entity_type, pagination, filters, file_url_func)
  File "/var/task/azul/service/repository.py", line 39, in _get_data
    entity_type=entity_type)
  File "/var/task/azul/service/responseobjects/elastic_request_builder.py", line 526, in transform_request
    final_response = FileSearchResponse(hits, paging, facets, entity_type)
  File "/var/task/azul/service/responseobjects/hca_response_v5.py", line 990, in __init__
    KeywordSearchResponse.__init__(self, hits, entity_type)
  File "/var/task/azul/service/responseobjects/hca_response_v5.py", line 898, in __init__
    self.map_entries(x) for x in hits], 'pagination': None}
  File "/var/task/azul/service/responseobjects/hca_response_v5.py", line 898, in <listcomp>
    self.map_entries(x) for x in hits], 'pagination': None}
  File "/var/task/azul/service/responseobjects/hca_response_v5.py", line 882, in map_entries
    donorOrganisms=self.make_donors(entry),
  File "/var/task/azul/service/responseobjects/hca_response_v5.py", line 827, in make_donors
    return [self.make_donor(donor) for donor in entry["contents"]["donors"]]
  File "/var/task/azul/service/responseobjects/hca_response_v5.py", line 827, in <listcomp>
    return [self.make_donor(donor) for donor in entry["contents"]["donors"]]
  File "/var/task/azul/service/responseobjects/hca_response_v5.py", line 817, in make_donor
    "donorCount": donor["donor_count"],
KeyError: 'donor_count'
```

So while it is a good default approach to use `[]` for keys that you expect to be present, this creates a problem during the integration test which hits the service endpoint *before* a reindexing can be triggered in Gitlab manually. I think we should use `.get` as for the other document fields.

**Comment by hannes-ucsc** [comment]
This requires re-indexing.

**Comment by hannes-ucsc** [comment]
This will require reindexing.

File: src/azul/service/responseobjects/hca_response_v5.py (hunk 5 above)
Comment (noah-aviel-dove) [reply src/azul/service/responseobjects/hca_response_v5.py]: Ah, so in other words, this function needs to process donors that were assembled by older versions of the code that don't have the other changes I made. Got it.

Comment (hannes-ucsc) [reply src/azul/service/responseobjects/hca_response_v5.py]: Yes, there is a little bit of a chicken-egg problem. We don't want to reindex until we've vetted the code but in this case the vetting fails because we haven't reindex yet. I'm currently unsure as to how we'll solve this.

Comment (hannes-ucsc) [reply src/azul/service/responseobjects/hca_response_v5.py]: I would like the service code to be strict when it comes to consuming the index, but at the moment that's not really possible.
